### PR TITLE
Dp sgd optimesers

### PR DESCRIPTION
**Pull request by MJHutchinson**

### CODE REVIEW

File: src/privacy_accounting/optimizer/dp_optimizer.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        microbatch_size = 1 if self.num_microbatches is None else ceil(loss.shape[0] / self.num_microbatches)
+
+        microbatches_losses = loss.split(microbatch_size, dim=0)
+
+        def process_microbatch(losses, sample_state):
+            self.optimiser.zero_grad()
+            microbatch_loss = losses.mean(dim=0)
+            microbatch_loss.backward(retain_graph=True)
+            record = self.get_grads(param_groups)
+            sample_state = self.dp_sum_query.accumulate_record(sample_params, sample_state, record)
```
Comment (MrinankSharma) [inline src/privacy_accounting/optimizer/dp_optimizer.py]: not 100% sure the methodology behind this - it seems really convoluted to me?

File: src/privacy_accounting/optimizer/dp_optimizer.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            return sample_state
+
+        for losses in microbatches_losses:
+            sample_state = process_microbatch(losses, sample_state)
```
Comment (MrinankSharma) [inline src/privacy_accounting/optimizer/dp_optimizer.py]: add comment - also doing accumulation

File: src/privacy_accounting/analysis/privacy_ledger.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        the datashard being kept private by the query. I.e. The amount of data
+        used to train in each epoch.
+        :param selection_probability: A (variable) float specifying the probability
+        that each record in the datashard is included in any given sample, i.e. the
+        number of minibatches.
+        """
+        self._population_size = population_size
+        self._selection_probability = selection_probability
+
+        init_capacity = ceil(1 / self._selection_probability)
```
[the remark this answers is not in the document]
Comment (MJHutchinson) [reply src/privacy_accounting/analysis/privacy_ledger.py]: Its just an initial guess, and means the buffer is always 2^n * epoch_size

File: src/privacy_accounting/dp_query/dp_query.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
+import abc
+
+import torch
+
+import src.utils.torch_nest_utils as nest
+
+
+class DPQuery(object):
+    __metaclass__ = abc.ABCMeta
```
[the remark this answers is not in the document]
Comment (MJHutchinson) [reply src/privacy_accounting/dp_query/dp_query.py]: If we had another mechanism (say Laplace noise, then its helpful to have an interface).

[the remark this answers is not in the document]
Comment (MJHutchinson) [reply src/privacy_accounting/dp_query/dp_query.py]: Technically is yeah, shouldnt make a difference but will change

[the remark this answers is not in the document]
Comment (MJHutchinson) [reply src/privacy_accounting/dp_query/dp_query.py]: https://arxiv.org/pdf/1808.10410.pdf
Laplace noise paper

File: src/privacy_accounting/dp_query/dp_query.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+        del ledger
+        raise TypeError(
+            'DPQuery typ %s does not support set_ledger.' % type(self).__name__
+        )
+
+    def initial_global_state(self):
+        """ Initial state for the global settings of the DPQuery. """
+
+        return ()
```
[the remark this answers is not in the document]
Comment (MJHutchinson) [reply src/privacy_accounting/dp_query/dp_query.py]: Again, just an interface, so any query should have its own initial global state, L2 clip and noise var for Gaussian mechanism

File: tests/test_PVI.py [hunk 6]
```diff
@@ -65,8 +67,8 @@ def no_privacy_function(update):
     all_y = np.concatenate([data['y'] for data in data_shards], axis=0)
 
     exact_inference_parameters = {
-        'slope_eta_1': (all_x.T @ all_x) / (1/stats[3]),
-        'slope_eta_2': (all_x.T @ all_y) / (1/stats[3])
+        'slope_eta_1': (all_x.T @ all_x) / (1 / stats[3]),
```
[the remark this answers is not in the document]
Comment (MJHutchinson) [reply tests/test_PVI.py]: Effectively has an uninformative prior rn, can change

File: tests/dp_optimiser_test.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+)
+batch_size = 10
+dataloader = data.DataLoader(dataset, batch_size=batch_size)
+
+optimiser = torch.optim.SGD(model.parameters(), lr=0.001)
+ledger = PrivacyLedger(x.shape[0], batch_size / x.shape[0])
+
+dp_optimiser = DPGaussianOptimiser(
+    l2_norm_clip=5,
+    noise_multiplier=1,
```
[the remark this answers is not in the document]
Comment (MJHutchinson) [reply tests/dp_optimiser_test.py]: So in the underlying code, everything works in an absolute value for the noise. Since you wanted to work in SNR, this just calculates the correct noise scale from the l2 clipping bound. I can rename if thats clearer

File: src/utils/torch_tensor_buffer.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
+import torch
+
+class TensorBuffer(object):
+    """ Lightweight class to hold series fo tensors of a given shape
+    in a size adaptive tensor. Utilised via the append method.
+    """
+
+    def __init__(self, capacity, shape, dtype=torch.float32):
```
[the remark this answers is not in the document]
Comment (MJHutchinson) [reply src/utils/torch_tensor_buffer.py]: float32 should be fine, 64 is double precision.

[the remark this answers is not in the document]
Comment (MJHutchinson) [reply src/utils/torch_tensor_buffer.py]: as a double in C

[the remark this answers is not in the document]
Comment (MJHutchinson) [reply src/utils/torch_tensor_buffer.py]: or Java

File: src/privacy_accounting/optimizer/wrapper_optimiser.py [hunk 9]
```diff
@@ -0,0 +1,7 @@
+import abc
+
+
+class WrapperOptimiser(object):
```
[the remark this answers is not in the document]
Comment (MJHutchinson) [reply src/privacy_accounting/optimizer/wrapper_optimiser.py]: Yep

File: src/privacy_accounting/dp_query/dp_query.py [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+        :param global_parameters: The current global parameters
+        :return: A tuple (results, new_global_state) that contains the result of the
+        query and the new query global state
+        """
+
+
+class SumAggregationDPQuery(DPQuery):
+
+    def initial_sample_state(self, param_groups):
+        return nest.map_structure(torch.zeros_like, param_groups)
```
[the remark this answers is not in the document]
Comment (MJHutchinson) [reply src/privacy_accounting/dp_query/dp_query.py]: Exactly what it is doing yeh

File: src/privacy_accounting/dp_query/dp_query.py [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+
+    def accumulate_preprocesses_record(self, sample_state, record):
```
[the remark this answers is not in the document]
Comment (MJHutchinson) [reply src/privacy_accounting/dp_query/dp_query.py]: Initial sample state is all zeros, as you dont want any gradients to begin with. The accumulate records function here just adds a new clipped gradient to the current stack of clipped gradients

File: src/privacy_accounting/dp_query/gaussian_query.py [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    def make_global_state(self, l2_norm_clip, noise_stddev):
+        return self._GlobalState(
+            torch.Tensor([l2_norm_clip]),
+            torch.Tensor([noise_stddev])
+        )
+
+    def initial_global_state(self):
+        return self.make_global_state(self._l2_norm_clip, self._noise_stddev)
+
+    def derive_sample_params(self, global_state):
```
[the remark this answers is not in the document]
Comment (MJHutchinson) [reply src/privacy_accounting/dp_query/gaussian_query.py]: Sample parameters describe how you should do things to this particular sample from the dataset. I.e. in the Gaussian mechanism case, what L2 clipping bound you should apply to the gradients. Passing it around like this gives the option for adapting these parameters during training

File: src/privacy_accounting/dp_query/gaussian_query.py [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+        return global_state.l2_norm_clip
+
+    def initial_sample_state(self, param_groups):
+        return nest.map_structure(torch.zeros_like, param_groups)
+
+    def preprocess_record(self, params, record):
+        l2_norm_clip = params
```
[the remark this answers is not in the document]
Comment (MJHutchinson) [reply src/privacy_accounting/dp_query/gaussian_query.py]: It may have changed if we're doing adaptive things

File: src/privacy_accounting/optimizer/dp_optimizer.py [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+    def __init__(self,
+                 optimiser,
+                 model,
+                 vector_loss,
+                 dp_sum_query,
+                 num_microbatches=None):
+
+        self.optimiser = optimiser
+        self.model = model
+        self.vector_loss = vector_loss
```
[the remark this answers is not in the document]
Comment (MJHutchinson) [reply src/privacy_accounting/optimizer/dp_optimizer.py]: Can do it its clearer

[the remark this answers is not in the document]
Comment (MJHutchinson) [reply src/privacy_accounting/optimizer/dp_optimizer.py]: *if

File: src/privacy_accounting/optimizer/dp_optimizer.py [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+        self.dp_sum_query = dp_sum_query
+        self.num_microbatches = num_microbatches
+
+        self._global_parameters = self.dp_sum_query.initial_global_state()
+
+    def fit_batch(self, x: torch.Tensor, y: torch.Tensor):
+
+        loss = self.vector_loss(self.model(x), y)
```
[the remark this answers is not in the document]
Comment (MJHutchinson) [reply src/privacy_accounting/optimizer/dp_optimizer.py]: yes - nn.module overrides the class __call__ function so that when doing that ^ the __call__ function is called, does some logic with hooks, calls the .forward function and the cleans up

[the remark this answers is not in the document]
Comment (MJHutchinson) [reply src/privacy_accounting/optimizer/dp_optimizer.py]: Go check nn.Module.__call__ for more details

File: src/privacy_accounting/optimizer/dp_optimizer.py [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+
+        param_groups = self.optimiser.param_groups
+
+        sample_state = self.dp_sum_query.initial_sample_state(
```
[the remark this answers is not in the document]
Comment (MJHutchinson) [reply src/privacy_accounting/optimizer/dp_optimizer.py]: Basically the optimiser stores a reference to any parameter you want it to optimise. This grabs a reference to all those so that we can work on thier graidents

File: src/privacy_accounting/optimizer/dp_optimizer.py [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+        )
+        sample_params = self.dp_sum_query.derive_sample_params(self._global_parameters)
+
+        microbatch_size = 1 if self.num_microbatches is None else ceil(loss.shape[0] / self.num_microbatches)
+
+        microbatches_losses = loss.split(microbatch_size, dim=0)
+
+        def process_microbatch(losses, sample_state):
+            self.optimiser.zero_grad()
+            microbatch_loss = losses.mean(dim=0)
```
[the remark this answers is not in the document]
Comment (MJHutchinson) [reply src/privacy_accounting/optimizer/dp_optimizer.py]: Mean on the gradients contributing to a micorbatch I'm p sure, sum on the clipped microbatch gradients.

File: src/privacy_accounting/optimizer/dp_optimizer.py (hunk 1 above)
Comment (MJHutchinson) [reply src/privacy_accounting/optimizer/dp_optimizer.py]: This is the way you have to do it. 
1. get the autodiff engine to compute the grad for the specific microbatch loss
2. grab those gradients off the parameters to clip
3. accumulate those gradients with what we have already accumulated. (we cant store this on the parameter itself as the autodiff will overwrite it on the next backwards pass)
4. zero out the grads and more onto the next micro batch
5. When youve done all the microbatches, apply noise and put the gradients back on the parameters so the optimiser knows what to do
6. Step the optimeser

File: src/privacy_accounting/optimizer/dp_optimizer.py (hunk 2 above)
Comment (MJHutchinson) [reply src/privacy_accounting/optimizer/dp_optimizer.py]: done

File: src/privacy_accounting/analysis/privacy_ledger.py [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+    'GaussianSumQueryEntry', ['l2_norm_bound', 'noise_stddev'])
+
+
+def format_ledger(sample_array, query_array):
+    """Converts array representation into a list of SampleEntries."""
+    samples = []
+    query_pos = 0
+    sample_pos = 0
+    for sample in sample_array:
+        population_size, selection_probability, num_queries = sample
```
[the remark this answers is not in the document]
Comment (MJHutchinson) [reply src/privacy_accounting/analysis/privacy_ledger.py]: copy paste error from github, fixed

File: src/privacy_accounting/analysis/privacy_ledger.py [hunk 19]
```diff
@@ -0,0 +1,141 @@
...
+
+SampleEntry = collections.namedtuple(  # pylint: disable=invalid-name
+    'SampleEntry', ['population_size', 'selection_probability', 'queries'])
+
+GaussianSumQueryEntry = collections.namedtuple(  # pylint: disable=invalid-name
+    'GaussianSumQueryEntry', ['l2_norm_bound', 'noise_stddev'])
+
+
+def format_ledger(sample_array, query_array):
+    """Converts array representation into a list of SampleEntries."""
```
[the remark this answers is not in the document]
Comment (MJHutchinson) [reply src/privacy_accounting/analysis/privacy_ledger.py]: sure can do in comments

File: src/privacy_accounting/analysis/privacy_ledger.py [hunk 20]
```diff
@@ -0,0 +1,141 @@
...
+        queries.append(GaussianSumQueryEntry(*query[1:]))
+        query_pos += 1
+    samples.append(SampleEntry(population_size, selection_probability, queries))
+    sample_pos += 1
+    return samples
+
+
+class PrivacyLedger(object):
+    """ Class for keeping a record of all the privacy events that occur
+    through a DPQuery executed on a given dataset.
```
[the remark this answers is not in the document]
Comment (MJHutchinson) [reply src/privacy_accounting/analysis/privacy_ledger.py]: You could apply it to either, it will depend on the query being applied

File: src/privacy_accounting/optimizer/dp_optimizer.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (MrinankSharma) [reply src/privacy_accounting/optimizer/dp_optimizer.py]: what does the sample state actually store?

File: src/privacy_accounting/optimizer/dp_optimizer.py (hunk 1 above)
Comment (MrinankSharma) [reply src/privacy_accounting/optimizer/dp_optimizer.py]: could we just comment what the sample params and state are? just to be clear - the sample params are the parameters for the DP mechanism ? state the current accumulated gradients?

File: src/privacy_accounting/optimizer/wrapper_optimiser.py [hunk 21]
```diff
@@ -0,0 +1,7 @@
+import abc
+
+
+class WrapperOptimiser(abc.ABCMeta):
```
Comment (MrinankSharma) [inline src/privacy_accounting/optimizer/wrapper_optimiser.py]: should this extent `ABC` ?

File: src/privacy_accounting/optimizer/dp_optimizer.py (hunk 1 above)
Comment (MJHutchinson) [reply src/privacy_accounting/optimizer/dp_optimizer.py]: Yeah will work on terminology throughout, inherited a bit from the TF lib I dont like as well

File: src/privacy_accounting/analysis/privacy_ledger.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (MJHutchinson) [reply src/privacy_accounting/analysis/privacy_ledger.py]: Will do

File: src/privacy_accounting/dp_query/dp_query.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (MJHutchinson) [reply src/privacy_accounting/dp_query/dp_query.py]: You dont have to override this method. Because all of this type should return some kind of tuple, its better to just pass an empty tuple rather than require a downstream class to check for None

File: tests/dp_optimiser_test.py [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+    optimiser=optimiser,
+    model=model,
+    ledger=ledger,
+    vector_loss=vec_loss,
+    num_microbatches=None
```
[the remark this answers is not in the document]
Comment (MJHutchinson) [reply tests/dp_optimiser_test.py]: Because this is a scrip for me to test things with, and in a demonstration its better if things are explicitly laid out imo

File: src/privacy_accounting/dp_query/dp_query.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (MJHutchinson) [reply src/privacy_accounting/dp_query/dp_query.py]: Have done

File: src/privacy_accounting/optimizer/dp_optimizer.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (MJHutchinson) [reply src/privacy_accounting/optimizer/dp_optimizer.py]: The current state of the sample being drawn from the dataset. So in microbatch SGD, the accumulated clipped gradients so far

File: src/privacy_accounting/optimizer/dp_optimizer.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (MJHutchinson) [reply src/privacy_accounting/optimizer/dp_optimizer.py]: This is the proceduce laid out in https://arxiv.org/pdf/1812.06210.pdf for micro-batching, I think it is correct. We use micro-batching to create better gradient estimates before clipping, hence we average rather than sum. When setting learning rates, this could be something to bear in mind youre right.

File: src/privacy_accounting/dp_query/dp_query.py [hunk 23]
```diff
@@ -0,0 +1,114 @@
+import abc
+
+import torch
+
+import src.utils.torch_nest_utils as nest
+
+
+class DPQuery(abc.ABCMeta):
```
[the remark this answers is not in the document]
Comment (MJHutchinson) [reply src/privacy_accounting/dp_query/dp_query.py]: Yes

File: src/privacy_accounting/optimizer/dp_optimizer.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (MrinankSharma) [reply src/privacy_accounting/optimizer/dp_optimizer.py]: add a comment pls
